### PR TITLE
Monitor refactor to support grafana multiple tenants

### PR DESCRIPTION
**Pull request by aiwantaozi**
1. update to use go mod
2. support get namespace metrics by rancher user name

### CODE REVIEW

File: Dockerfile.dapper [hunk 1]
```diff
@@ -1,19 +1,12 @@
-FROM golang:1.11.1-alpine3.8
+FROM golang:1.13.10-alpine3.11
 
+ENV GOPROXY=https://goproxy.cn,direct
```
Comment (thxCode) [inline Dockerfile.dapper]: `GOPROXY` is not needed if we configure `GO111MODULE=off` below.

File: Dockerfile.dapper [hunk 2]
```diff
@@ -1,19 +1,12 @@
...
-    rm -rf /go/src /go/pkg
-
-RUN go get github.com/rancher/trash && \
-    go get -d github.com/alecthomas/gometalinter && \
-    git -C /go/src/github.com/alecthomas/gometalinter checkout -b current v2.0.11 && \
-    go install github.com/alecthomas/gometalinter && \
-    gometalinter --install && \
-    rm -rf /go/src /go/pkg
-
+RUN if [ "${ARCH}" == "amd64" ]; then \
```
Comment (thxCode) [inline Dockerfile.dapper]: Replace "${ARCH}" with "$(go env GOARCH)"

File: scripts/validate [hunk 3]
```diff
@@ -9,12 +9,17 @@ PACKAGES="$(go list ./...)"
...
 go vet ${PACKAGES}
-echo Running: gometalinter
-for i in ${PACKAGES}; do
-    if [ -n "$(gometalinter $i | grep -v 'should have comment.*or be unexported' | tee /dev/stderr)" ]; then
-        failed=true
-    fi
-done
-test -z "$failed"
+
+if ! command -v golangci-lint; then
```
Comment (thxCode) [inline scripts/validate:10]: If not found `golangci-lint`, it's better to use `go fmt/vet` instead.

File: scripts/validate [hunk 4]
```diff
@@ -9,12 +9,17 @@ PACKAGES="$(go list ./...)"
...
+    echo Skipping validation: no golangci-lint available
+    exit
+else
+    echo Running: golangci-lint
+    golangci-lint run --skip-files "zz_.*" --skip-files "pkg/agent/grpc.go"
```
Comment (thxCode) [inline scripts/validate]: `golang-lint` supports configuration model, we can use `.golangci.json` instead.

File: scripts/validate [hunk 5]
```diff
@@ -9,12 +9,17 @@ PACKAGES="$(go list ./...)"
...
+fi
+
 echo Running: go fmt
 test -z "$(go fmt ${PACKAGES} | tee /dev/stderr)"
```
Comment (thxCode) [inline scripts/validate]: golangci-lint can do `gofmt` if it's available.

File: scripts/validate [hunk 6]
```diff
@@ -9,12 +9,17 @@ PACKAGES="$(go list ./...)"
...
+
+echo Running: go mod verify
+go mod verify
```
Comment (thxCode) [inline scripts/validate]: Firstly, there is not any module under `GOPATH` if `GO111MODULE=off`, so we should give a condition in here. 
```
if [[ "$(go env GO111MODULE)" == "off" ]]; then
   ## warning
else
  ## go mod ...
fi
```

Secondly, `go mod verify` only check the cached dependencies, it will not verify the security of the package, and then it cannot clean up the `vendor` dir, which will cause a dirty tree. Please change to `go mod tidy; go mod vendor`.

File: pkg/agent/const.go [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package agent
+
+const (
+	contentTypeHeader     = "Content-Type"
+	contentEncodingHeader = "Content-Encoding"
+	acceptHeader          = "Accept"
+	jsonContentType       = "application/json"
+	protoContentType      = "application/x-protobuf"
```
Comment (thxCode) [inline pkg/agent/const.go:8]: In fact, we cannot see any benefit to extract these fields into a single const file, why we don't keep to use cockroachdb's constants?

File: pkg/agent/entry.go [hunk 8]
```diff
@@ -202,3 +244,17 @@ func (a *agent) createGRPCProxy() *grpc.Server {
...
+	kubeConfig := os.Getenv("KUBECONFIG")
+	if kubeConfig != "" {
+		cfg, err := getExternal(kubeConfig)
+		return cfg, err
+	}
+
+	return rest.InClusterConfig()
+}
+
+func getExternal(kubeConfig string) (*rest.Config, error) {
```
Comment (thxCode) [inline pkg/agent/entry.go]: `getExternal` function can be inlined into `getKubeConfig` directly, there is not another caller.

File: pkg/agent/http.go [hunk 9]
```diff
@@ -59,16 +62,55 @@ func accessControl(agt *agent, proxyHandler http.Handler) http.Handler {
...
-				return
+			var namespaceSet data.Set
+			if rancherUser != "" || len(rancherGroup) != 0 {
+				info := &user.DefaultInfo{
+					Name:   rancherUser,
+					UID:    rancherUser,
+					Groups: rancherGroup,
+				}
+
+				log.Debugf("%s - %s - user info: %+v", r.Method, r.URL.Path, info)
```
Comment (thxCode) [inline pkg/agent/http.go]: I think is not a good idea to print the user information.

File: pkg/agent/http.go [hunk 10]
```diff
@@ -59,16 +62,55 @@ func accessControl(agt *agent, proxyHandler http.Handler) http.Handler {
...
+
+				if agt.nodes.CanList(info) {
+					proxyHandler.ServeHTTP(w, r)
+					return
+				}
+
+				namespaceSet = agt.namespaces.QueryByUser(info)
+			} else if len(accessToken) != 0 {
+				log.Debugf("%s - %s - accessToken info: %+v", r.Method, r.URL.Path, accessToken)
```
Comment (thxCode) [inline pkg/agent/http.go]: I think is not a good idea to print the access token information.

File: pkg/kube/nodes.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"context"
+
+	"github.com/rancher/prometheus-auth/pkg/kube/rbac"
+	"github.com/rancher/steve/pkg/accesscontrol"
+	v1 "github.com/rancher/types/apis/core/v1"
+
+	k8scorev1 "k8s.io/api/core/v1"
+	schema2 "k8s.io/apimachinery/pkg/runtime/schema"
```
Comment (thxCode) [inline pkg/kube/nodes.go]: `schema2` is not a readable package alias.

File: pkg/kube/rbac/policy_rule_index.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+	"sort"
+
+	"github.com/rancher/steve/pkg/accesscontrol"
+	v1 "github.com/rancher/wrangler-api/pkg/generated/controllers/rbac/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	All       = "*"
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: go format? and why the `All` constant is publicly visible?

File: pkg/kube/rbac/policy_rule_index.go [hunk 13]
```diff
@@ -0,0 +1,171 @@
+package rbac
+
+import (
+	"hash"
+	"sort"
+
+	"github.com/rancher/steve/pkg/accesscontrol"
+	v1 "github.com/rancher/wrangler-api/pkg/generated/controllers/rbac/v1"
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: Doesn't need an alias.

File: pkg/kube/rbac/policy_rule_index.go [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	for _, rb := range p.getRoleBindings(subjectName) {
+		switch rb.RoleRef.Kind {
+		case "Role":
+			digest.Write([]byte(rb.RoleRef.Name))
+			digest.Write([]byte(rb.Namespace))
+			digest.Write([]byte(p.revisions.roleRevision(rb.Namespace, rb.RoleRef.Name)))
+			digest.Write(null)
+		case "ClusterRole":
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: Sorry, I cannot see what's the difference processing between `Role` and `ClusterRole`, why we need a `switch case` block here?

File: pkg/kube/rbac/policy_rule_index.go [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+const (
+	All       = "*"
+)
+
+type policyRuleIndex struct {
+	crCache             v1.ClusterRoleCache
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: The short name field in a struct is not a good pattern, we should keep in the readable full name.

File: pkg/kube/rbac/sa_access.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
+package rbac
+
+import (
+"context"
+"crypto/sha256"
+"encoding/hex"
+"time"
+
+v1 "github.com/rancher/wrangler-api/pkg/generated/controllers/rbac/v1"
+"k8s.io/apimachinery/pkg/util/cache"
```
Comment (thxCode) [inline pkg/kube/rbac/sa_access.go]: go format the imports.

File: pkg/kube/rbac/policy_rule_index.go [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+func newSAPolicyRuleIndex(revisions *roleRevisionIndex, rbac v1.Interface) *policyRuleIndex {
+	key := "ServiceAccount"
+
+	pi := &policyRuleIndex{
+		kind:                key,
+		crCache:             rbac.ClusterRole().Cache(),
+		rCache:              rbac.Role().Cache(),
+		crbCache:            rbac.ClusterRoleBinding().Cache(),
+		rbCache:             rbac.RoleBinding().Cache(),
+		clusterRoleIndexKey: "crb" + key,
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: Don't combine an indexer key like this, just simply use `crbServiceAccount` or `IndexByServiceAccountSubject`.

File: pkg/kube/rbac/policy_rule_index.go [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+		roleIndexKey:        "rb" + key,
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: Doesn't combine an indexer key like this, just simply use `rbServiceAccount` or `IndexByServiceAccountSubject`.

File: pkg/kube/rbac/policy_rule_index.go [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+	All       = "*"
+)
+
+type policyRuleIndex struct {
+	crCache             v1.ClusterRoleCache
+	rCache              v1.RoleCache
+	crbCache            v1.ClusterRoleBindingCache
+	rbCache             v1.RoleBindingCache
+	revisions           *roleRevisionIndex
+	kind                string
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: We can just compare `subject.Kind == "ServiceAccount"` below as I cannot find any reason why we should use a field to identify the kind of this indexer.

File: pkg/kube/rbac/policy_rule_index.go [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+
+	pi.crbCache.AddIndexer(pi.clusterRoleIndexKey, pi.clusterRoleBindingBySubjectIndexer)
+	pi.rbCache.AddIndexer(pi.roleIndexKey, pi.roleBindingBySubject)
+
+	return pi
+}
+
+func (p *policyRuleIndex) clusterRoleBindingBySubjectIndexer(crb *rbacv1.ClusterRoleBinding) (result []string, err error) {
+	for _, subject := range crb.Subjects {
+		if subject.Kind == p.kind && crb.RoleRef.Kind == "ClusterRole" {
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: Why we need to compare `crb.RoleRef.Kind == "ClusterRole"` in here?

File: pkg/kube/rbac/sa_access.go [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+	return as
+}
+
+func (l *AccessStore) AccessFor(sa *k8scorev1.ServiceAccount) *accesscontrol.AccessSet {
+	var cacheKey string
+	if l.cache != nil {
+		cacheKey = l.CacheKey(sa)
+		val, ok := l.cache.Get(cacheKey)
+		if ok {
+			as, _ := val.(*accesscontrol.AccessSet)
```
Comment (thxCode) [inline pkg/kube/rbac/sa_access.go]: Remove the second receiver(`_`) if we confirm the type is `*accesscontrol.AccessSet`.

File: pkg/agent/entry.go [hunk 22]
```diff
@@ -181,11 +184,50 @@ func createAgent(cfg *agentConfig) (*agent, error) {
...
+
+	controllerFactory, err := controller.NewSharedControllerFactoryFromConfig(k8sConfig, scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	coreClient := core.New(controllerFactory)
+	rbacClient := rbac.New(controllerFactory)
+	userAccessStore := accesscontrol.NewAccessStore(cfg.ctx, false, rbacClient.V1())
+	saAccessStore := kuberbac.NewSAAccessStore(cfg.ctx, false, rbacClient.V1())
```
Comment (thxCode) [inline pkg/agent/entry.go]: So, if we don't want to cache the ServiceAcount, why we implement a cache?

File: pkg/kube/rbac/role_revision_index.go [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+		namespace: namespace,
+	})
+	revision, _ := val.(string)
+	return revision
+}
+
+func (r *roleRevisionIndex) onClusterRoleChanged(key string, cr *rbacv1.ClusterRole) (role *rbacv1.ClusterRole, err error) {
+	if cr == nil {
+		r.roleRevisions.Delete(roleKey{
+			name: key,
```
Comment (thxCode) [inline pkg/kube/rbac/role_revision_index.go]: I don't understand the `OnChange` callback, the `key` input parameter is a namespaced name like `default/xxx`? If it's what I guessed, why we don't cache the namespaced name directly?

File: pkg/kube/rbac/role_revision_index.go [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+
+func newRoleRevision(ctx context.Context, rbac rbac.Interface) *roleRevisionIndex {
+	r := &roleRevisionIndex{}
+	rbac.Role().OnChange(ctx, "role-revision-indexer", r.onRoleChanged)
+	rbac.ClusterRole().OnChange(ctx, "role-revision-indexer", r.onClusterRoleChanged)
+	return r
+}
+
+func (r *roleRevisionIndex) roleRevision(namespace, name string) string {
+	val, _ := r.roleRevisions.Load(roleKey{
```
Comment (thxCode) [inline pkg/kube/rbac/role_revision_index.go]: Why we don't use the namespaced name instead of `roleKey` directly?

File: pkg/kube/rbac/role_revision_index.go [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+	"context"
+	"sync"
+
+	rbac "github.com/rancher/wrangler-api/pkg/generated/controllers/rbac/v1"
+	"github.com/rancher/wrangler/pkg/kv"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+type roleRevisionIndex struct {
+	roleRevisions sync.Map
```
Comment (thxCode) [inline pkg/kube/rbac/role_revision_index.go]: It's better to name `indexer`.

File: pkg/kube/rbac/policy_rule_index.go [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	for _, binding := range p.getClusterRoleBindings(subjectName) {
+		p.addAccess(result, All, binding.RoleRef)
+	}
+
+	return result
+}
+
+func (p *policyRuleIndex) addAccess(accessSet *accesscontrol.AccessSet, namespace string, roleRef rbacv1.RoleRef) {
```
Comment (thxCode) [inline pkg/kube/rbac/policy_rule_index.go]: There are five `for` blocks in its implementation, can we separate them?

File: scripts/validate (hunk 3 above)
Comment (thxCode) [reply scripts/validate:10]: https://github.com/cnrancher/octopus/blob/26235beed387f9f0bc0e74d4659c128711756ed0/hack/lib/lint.sh#L28-L46

File: Dockerfile.dapper (hunk 1 above)
Comment (aiwantaozi) [reply Dockerfile.dapper]: updated

File: Dockerfile.dapper (hunk 2 above)
Comment (aiwantaozi) [reply Dockerfile.dapper]: updated

File: scripts/validate (hunk 4 above)
Comment (aiwantaozi) [reply scripts/validate]: updated

File: pkg/agent/entry.go (hunk 22 above)
Comment (aiwantaozi) [reply pkg/agent/entry.go]: updated

File: pkg/kube/rbac/role_revision_index.go (hunk 23 above)
Comment (aiwantaozi) [reply pkg/kube/rbac/role_revision_index.go]: prefer to keep it as steve do, so we can distinguish clusterRole and role when we need.

File: pkg/kube/rbac/role_revision_index.go (hunk 25 above)
Comment (aiwantaozi) [reply pkg/kube/rbac/role_revision_index.go]: updated

File: pkg/kube/rbac/policy_rule_index.go (hunk 26 above)
Comment (aiwantaozi) [reply pkg/kube/rbac/policy_rule_index.go]: updated

File: pkg/agent/const.go (hunk 7 above)
Comment (aiwantaozi) [reply pkg/agent/const.go:8]: Replace these fields with const, then we can remove package github.com/cockroachdb/cockroach/pkg/util/httputil

File: pkg/kube/rbac/policy_rule_index.go (hunk 13 above)
Comment (aiwantaozi) [reply pkg/kube/rbac/policy_rule_index.go]: it automatically added by goimports

File: pkg/kube/nodes.go (hunk 11 above)
Comment (aiwantaozi) [reply pkg/kube/nodes.go]: updated

File: pkg/kube/rbac/policy_rule_index.go (hunk 15 above)
Comment (aiwantaozi) [reply pkg/kube/rbac/policy_rule_index.go]: updated

File: pkg/kube/rbac/policy_rule_index.go (hunk 19 above)
Comment (aiwantaozi) [reply pkg/kube/rbac/policy_rule_index.go]: updated

File: pkg/kube/rbac/policy_rule_index.go (hunk 18 above)
Comment (aiwantaozi) [reply pkg/kube/rbac/policy_rule_index.go]: updated

File: pkg/kube/rbac/sa_access.go (hunk 21 above)
Comment (aiwantaozi) [reply pkg/kube/rbac/sa_access.go]: updated

**Comment by thxCode** [comment]
~LGTM~ For vendor dirty tree verification, we need to discuss further.

File: scripts/validate (hunk 6 above)
Comment (orangedeng) [reply scripts/validate]: `go mod vendor` will cause vendor recreate, I will cause unnecessary vendor download when running CI.
It is good enough that running `go mod verify` and the CI process will be stopped if the tree is not clean.

Comment (thxCode) [reply scripts/validate]: With `export GO111MODULE=off` indicated in Dockerfile.dapper, the `go mod` operation will be broken, therefore, we give a judgment in here. 

- When executing `make validate`, `go mod` will be skipped. _An incomplete vendor will cause the build to fail and can also act as a reminder._
- When executing `./scripts/validate` in the laptop, `go mod tidy; go mod vendor` will help us to clean the vendor tree.

**Comment by thxCode** [comment]
A quick verification:

- go get github.com/gorilla/mux@master
- git commit the `go.mod`, `go.sum` files without vendor changes.
- based on this PR, modify the `validate.sh` to execute `go mod verify` forcibly: `GO111MODULE=on go mod verify`

The result is `all modules verified`, the `go mod verify` cannot find out the `github.com/gorilla/mux` vendor has changed to the master version. I think the `go mod verify` only verifies the project under `go mod` module.

We use the vendor directory for caching the dependencies of building, so we put our project under GOPATH and `export GO111MODULE=off`. However, if we forget to update our vendor directory(i.e, update the version of an existing vendor, like what I did above), there's not any way to check our vendor's state. 

> go mod verify does not download content for modules not in the cache, and it does not use go.sum files to verify module content. 

Combine [the description](https://golang.org/ref/mod#tmp_33) of official doc and the comment of [go issue#36310](https://github.com/golang/go/issues/36130#issuecomment-566115897,) the `go mod verify` is to detect tampering with the cache after modules are downloaded, but not to verify the vendor tree.

In short, the vendor is absolutely not a module shadow. My suggestion is as below:

- If 100% go mod management, we can use a Docker multistage-build to cache the modules like what Kubebuilder does: https://github.com/kubernetes-sigs/kubebuilder/blob/26e468a3aa1c91e20b811eb763565b68c5576410/pkg/plugin/v3/scaffolds/internal/templates/dockerfile.go#L45-L50
- If 100% based on vendor, we need a convention to tell contributors: `hey, you should make sure your vendor is as expected` or request reviewers to notice the change of `go.mod`. A possible way is to create a new CI step to verify the git tree's state after `GO111MODULE=on go mod tidy; GO111MODULE=on  go mod vendor`.

@aiwantaozi , cc @orangedeng

**Comment by orangedeng** [comment]
> A quick verification:
> 
> * go get github.com/gorilla/mux@master
> * git commit the `go.mod`, `go.sum` files without vendor changes.
> * based on this PR, modify the `validate.sh` to execute `go mod verify` forcibly: `GO111MODULE=on go mod verify`
> 
> The result is `all modules verified`, the `go mod verify` cannot find out the `github.com/gorilla/mux` vendor has changed to the master version. I think the `go mod verify` only verifies the project under `go mod` module.
> 
> We use the vendor directory for caching the dependencies of building, so we put our project under GOPATH and `export GO111MODULE=off`. However, if we forget to update our vendor directory(i.e, update the version of an existing vendor, like what I did above), there's not any way to check our vendor's state.
> 
> > go mod verify does not download content for modules not in the cache, and it does not use go.sum files to verify module content.
> 
> Combine [the description](https://golang.org/ref/mod#tmp_33) of official doc and the comment of [go issue#36310](https://github.com/golang/go/issues/36130#issuecomment-566115897,) the `go mod verify` is to detect tampering with the cache after modules are downloaded, but not to verify the vendor tree.
> 
> In short, the vendor is absolutely not a module shadow. My suggestion is as below:
> 
> * If 100% go mod management, we can use a Docker multistage-build to cache the modules like what Kubebuilder does: https://github.com/kubernetes-sigs/kubebuilder/blob/26e468a3aa1c91e20b811eb763565b68c5576410/pkg/plugin/v3/scaffolds/internal/templates/dockerfile.go#L45-L50
> * If 100% based on vendor, we need a convention to tell contributors: `hey, you should make sure your vendor is as expected` or request reviewers to notice the change of `go.mod`. A possible way is to create a new CI step to verify the git tree's state after `GO111MODULE=on go mod tidy; GO111MODULE=on  go mod vendor`.
> 
> @aiwantaozi , cc @orangedeng

I agree on the second way. Download vendor while building is not the ideal way to do CI process because we need to validate the vendor when summiting PRs. 
Please do `GO111MODULE=on go mod tidy; GO111MODULE=on  go mod vendor` in validate step.

File: pkg/agent/http.go [hunk 27]
```diff
@@ -59,16 +62,57 @@ func accessControl(agt *agent, proxyHandler http.Handler) http.Handler {
...
+			var namespaceSet data.Set
+			if rancherUser != "" || len(rancherGroup) != 0 {
+				log.Debugf("%s - %s - access by userID", r.Method, r.URL.Path)
+				info := &user.DefaultInfo{
+					Name:   rancherUser,
+					UID:    rancherUser,
+					Groups: rancherGroup,
+				}
+
+				if agt.nodes.CanList(info) {
```
Comment (orangedeng) [inline pkg/agent/http.go]: Code duplicated in Line 110-115. You can put them out of the if-else block.

File: pkg/kube/secrets.go [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+	ctx                  context.Context
+	secretCache          corev1.SecretCache
+	reviewResultTTLCache *cache.LRUExpireCache
+}
+
+func NewSecrets(ctx context.Context, secretCache corev1.SecretCache) *Secrets {
+	return &Secrets{
+		ctx:                  ctx,
+		secretCache:          secretCache,
+		reviewResultTTLCache: cache.NewLRUExpireCache(1024),
```
Comment (orangedeng) [inline pkg/kube/secrets.go]: Why do we need to have LRU cache here? I think we are using `secretCache` so we have all sa token cached in memory?

File: pkg/kube/namespaces.go [hunk 29]
```diff
@@ -2,182 +2,97 @@ package kube
...
+func NewNamespaces(ctx context.Context,
+	namespaceCache corev1.NamespaceCache, secrets *Secrets,
+	acl *accesscontrol.AccessStore, monitoringNs string) Namespaces {
+	return &namespaces{
+		ctx:                  ctx,
+		monitoringNamespace:  monitoringNs,
+		namespaceCache:       namespaceCache,
+		secrets:              secrets,
+		accessStore:          acl,
+		reviewResultTTLCache: cache.NewLRUExpireCache(1024),
```
Comment (orangedeng) [inline pkg/kube/namespaces.go]: Why we need `reviewResultTTLCache` here? I can't find any reference of this field.

File: vendor/github.com/rancher/steve/pkg/accesscontrol/policy_rule_index.go [hunk 30]
```diff
@@ -153,9 +153,7 @@ func (p *policyRuleIndex) getClusterRoleBindings(subjectName string) []*rbacv1.C
 	if err != nil {
 		return nil
 	}
-	sort.Slice(result, func(i, j int) bool {
-		return result[i].Name < result[j].Name
-	})
+	// call chains GetByIndex -> ByIndex -> IndexKeys (already sorted by name )
```
Comment (orangedeng) [inline vendor/github.com/rancher/steve/pkg/accesscontrol/policy_rule_index.go]: How can we ensure the the result sorted? Can you reference the code here?

**Comment by aiwantaozi** [comment]
> > A quick verification:
> > 
> > * go get github.com/gorilla/mux@master
> > * git commit the `go.mod`, `go.sum` files without vendor changes.
> > * based on this PR, modify the `validate.sh` to execute `go mod verify` forcibly: `GO111MODULE=on go mod verify`
> > 
> > The result is `all modules verified`, the `go mod verify` cannot find out the `github.com/gorilla/mux` vendor has changed to the master version. I think the `go mod verify` only verifies the project under `go mod` module.
> > We use the vendor directory for caching the dependencies of building, so we put our project under GOPATH and `export GO111MODULE=off`. However, if we forget to update our vendor directory(i.e, update the version of an existing vendor, like what I did above), there's not any way to check our vendor's state.
> > > go mod verify does not download content for modules not in the cache, and it does not use go.sum files to verify module content.
> > 
> > 
> > Combine [the description](https://golang.org/ref/mod#tmp_33) of official doc and the comment of [go issue#36310](https://github.com/golang/go/issues/36130#issuecomment-566115897,) the `go mod verify` is to detect tampering with the cache after modules are downloaded, but not to verify the vendor tree.
> > In short, the vendor is absolutely not a module shadow. My suggestion is as below:
> > 
> > * If 100% go mod management, we can use a Docker multistage-build to cache the modules like what Kubebuilder does: https://github.com/kubernetes-sigs/kubebuilder/blob/26e468a3aa1c91e20b811eb763565b68c5576410/pkg/plugin/v3/scaffolds/internal/templates/dockerfile.go#L45-L50
> > * If 100% based on vendor, we need a convention to tell contributors: `hey, you should make sure your vendor is as expected` or request reviewers to notice the change of `go.mod`. A possible way is to create a new CI step to verify the git tree's state after `GO111MODULE=on go mod tidy; GO111MODULE=on  go mod vendor`.
> > 
> > @aiwantaozi , cc @orangedeng
> 
> I agree on the second way. Download vendor while building is not the ideal way to do CI process because we need to validate the vendor when summiting PRs.
> Please do `GO111MODULE=on go mod tidy; GO111MODULE=on go mod vendor` in validate step.

@thxCode @orangedeng Yes, we do these in the validate step here. https://github.com/cnrancher/prometheus-auth/pull/1/commits/b75e0d1ea71b556f4895078bb282d20cbd99e97a#diff-fb91b8b530a699f09fb08e0dd2fdedbcR22

File: vendor/github.com/rancher/steve/pkg/accesscontrol/policy_rule_index.go (hunk 30 above)
Comment (aiwantaozi) [reply vendor/github.com/rancher/steve/pkg/accesscontrol/policy_rule_index.go]: We didn't sort the result in the wrangler-api implement, so we still need sort here.

https://github.com/aiwantaozi/prometheus-auth/blob/monitor-refactor/vendor/github.com/rancher/wrangler-api/pkg/generated/controllers/rbac/v1/clusterrolebinding.go#L246-L256

https://github.com/aiwantaozi/prometheus-auth/blob/monitor-refactor/vendor/k8s.io/client-go/tools/cache/thread_safe_store.go#L180-L198

File: pkg/kube/namespaces.go (hunk 29 above)
Comment (aiwantaozi) [reply pkg/kube/namespaces.go]: updated

File: pkg/kube/secrets.go (hunk 28 above)
Comment (aiwantaozi) [reply pkg/kube/secrets.go]: updated

File: pkg/agent/http.go (hunk 27 above)
Comment (aiwantaozi) [reply pkg/agent/http.go]: updated

**Comment by orangedeng** [comment]
LGTM

**Comment by thxCode** [comment]
LGTM,thanks
